crates/filament/src/ast_passes/unused_elements.rs

```diff
@@ -0,0 +1,421 @@
+use crate::ast_visitor::{Action, Visitor};
+use crate::cmdline;
+use fil_ast as ast;
+use fil_utils::{Diagnostics, Error, GPosIdx};
+use std::collections::{HashMap, HashSet};
+
+/// Structure to track usage information for a single component
+#[derive(Default, Clone)]
+struct ComponentUsage {
+    // Definitions with their locations
+    instances: HashMap<String, ast::Loc<fil_utils::Id>>,
+    invocations: HashMap<String, ast::Loc<fil_utils::Id>>,
+    let_params: HashMap<String, ast::Loc<fil_utils::Id>>,
+    sig_params: HashMap<String, ast::Loc<fil_utils::Id>>,
+
+    // Usage tracking
+    invoked_instances: HashSet<String>,
+    accessed_invocations: HashSet<String>,
+    referenced_params: HashSet<String>,
+}
```

Comment:
This approach tracks one single scope for the entire component but constructs like "for" and "if" create new scopes and bindings may shadow each other. This means that the single flat global scope would not work.